### PR TITLE
Resnet nhwc perf code

### PR DESCRIPTION
**Pull request by strint**
## NCHW

image shape  oneflow.Size([19, 3, 224, 224])

x shape after padding  oneflow.Size([19, 4, 224, 224])

x shape after conv1  oneflow.Size([19, 64, 112, 112])

x shape after bn1 oneflow.Size([19, 64, 112, 112])

x shape after maxpool  oneflow.Size([19, 64, 56, 56])

x shape after layer1  oneflow.Size([19, 256, 56, 56])

x shape after layer2  oneflow.Size([19, 512, 28, 28])

x shape after layer3  oneflow.Size([19, 1024, 14, 14])

x shape after layer4  oneflow.Size([19, 2048, 7, 7])

x shape after avgpool  oneflow.Size([19, 2048, 1, 1])

x shape after flatten  oneflow.Size([19, 2048])

x shape after fc  oneflow.Size([19, 1000])

## NHWC

image shape  oneflow.Size([19, 224, 224, 3])

x shape after padding  oneflow.Size([19, 224, 224, 4])

x shape after conv1  oneflow.Size([19, 112, 112, 64])

x shape after bn1 oneflow.Size([19, 112, 112, 64])

x shape after maxpool  oneflow.Size([19, 56, 56, 64])

x shape after layer1  oneflow.Size([19, 56, 56, 256])

x shape after layer2  oneflow.Size([19, 28, 28, 512])

x shape after layer3  oneflow.Size([19, 14, 14, 1024])

x shape after layer4  oneflow.Size([19, 7, 7, 2048])

x shape after avgpool  oneflow.Size([19, 1, 1, 2048])

x shape after flatten  oneflow.Size([19, 2048])

x shape after fc oneflow.Size([19, 1000])

### CODE REVIEW

File: Vision/classification/image/resnet50/models/resnet50.py [hunk 1]
```diff
@@ -206,12 +231,15 @@ def __init__(
         self.layer4 = self._make_layer(
             block, 512, layers[3], stride=2, dilate=replace_stride_with_dilation[2]
         )
-        self.avgpool = nn.AvgPool2d((7, 7))
+        if self.channel_last:
+           self.avgpool = nn.LegacyAvgPool2d((7, 7), data_format=self.data_format)
+        else:
+           self.avgpool = nn.AvgPool2d((7, 7))
```
Comment (strint) [inline Vision/classification/image/resnet50/models/resnet50.py:237]:
```
        if self.channel_last:
           self.avgpool = nn.LegacyAvgPool2d((7, 7), stride=(1, 1), data_format=self.data_format)
        else:
           self.avgpool = nn.AvgPool2d((7, 7), stride=(1, 1))
```
这里增加下stride，能对齐 pooling_bw_kernel_avg

File: Vision/classification/image/resnet50/models/data.py [hunk 2]
```diff
@@ -30,6 +30,7 @@ def make_data_loader(args, mode, is_consistent=False, synthetic=False):
             num_classes=args.num_classes,
             placement=placement,
             sbp=sbp,
+            channel_last=args.channels_last,
         )
```
Comment (ouyangyu) [inline Vision/classification/image/resnet50/models/data.py:34]: channel_last=args.channel_last
多了一个s

Comment (strint) [reply Vision/classification/image/resnet50/models/data.py:34]: fixed

File: Vision/classification/image/resnet50/graph.py [hunk 3]
```diff
@@ -40,6 +40,9 @@ def __init__(
         self.config.allow_fuse_add_to_output(True)
         self.config.allow_fuse_model_update_ops(True)
         self.config.enable_cudnn_conv_heuristic_search_algo(False)
+        self.world_size = flow.env.get_world_size()
+        if self.world_size / args.num_devices_per_node > 1:
```
Comment (Ldpe2G) [inline Vision/classification/image/resnet50/graph.py:44]: 这里意思是多机的情况下才会打开这个吗？

File: Vision/classification/image/resnet50/models/resnet50.py [hunk 4]
```diff
@@ -121,6 +122,9 @@ def __init__(
 
     def forward(self, x: Tensor) -> Tensor:
         identity = x
+        if self.downsample is not None:
+            identity = self.downsample(x)
```
Comment (Ldpe2G) [inline Vision/classification/image/resnet50/models/resnet50.py:126]: 这段代码挪到前面有什么原因吗

File: Vision/classification/image/resnet50/models/resnet50.py [hunk 5]
```diff
@@ -279,6 +294,14 @@ def _make_layer(
         return nn.Sequential(*layers)
 
     def _forward_impl(self, x: Tensor) -> Tensor:
+        if self.pad_input:
+            if self.channel_last:
+                # NHWC
+                paddings = (0, 1)
+            else:
+                # NCHW
```
Comment (Ldpe2G) [inline Vision/classification/image/resnet50/models/resnet50.py:302]: pad_input 为 true 只会是 nhwc 的情况？而且  NHWC 分支的  paddings = (0, 1) 为什么只有两个数字这个怎么解释

**Comment by strint** [comment]
本地试跑了一下：oneflow master可以跑通本分支的单卡和2卡；

File: Vision/classification/image/resnet50/graph.py (hunk 3 above)
Comment (strint) [reply Vision/classification/image/resnet50/graph.py:44]: 是的，单卡下试跑有用，多卡下没用。

试跑就是不用启发式搜索。是一个优化策略。

File: Vision/classification/image/resnet50/models/resnet50.py (hunk 4 above)
Comment (strint) [reply Vision/classification/image/resnet50/models/resnet50.py:126]: 有，放到原来的位置，会带来2%的性能拖累。原因参见：https://github.com/Oneflow-Inc/OneTeam/issues/840#issuecomment-994903466

model.layer2.0.downsample.0-conv2d_35需要在model.layer2.0.conv1-conv2d_30前面添加到图中，才能避免上面的问题。

File: Vision/classification/image/resnet50/models/resnet50.py (hunk 5 above)
Comment (strint) [reply Vision/classification/image/resnet50/models/resnet50.py:302]: 这种写法是torch的接口。
(a, 0, 0, 0, c, d)表示((a, 0), (0, 0), (c, d))
- 即倒数第一维按(c, d) pad，即前面加c，后面加d；
- 倒数第二维按(0, 0) pad，即不扩展
- 倒数第三维按(a, 0) pad，即前面扩展a

Ref: https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html

**Review by ouyangyu — APPROVED** [review]

File: Vision/classification/image/resnet50/models/resnet50.py (hunk 4 above)
Comment (Ldpe2G) [reply Vision/classification/image/resnet50/models/resnet50.py:129]: 好的，写模块的顺序还会影响到 graph  的性能，可是实际用户怎么写是没办法遇见的

File: Vision/classification/image/resnet50/models/resnet50.py (hunk 5 above)
Comment (Ldpe2G) [reply Vision/classification/image/resnet50/models/resnet50.py:304]: 好的

**Review by Ldpe2G — APPROVED** [review]
